storage-proofs/porep/src/stacked/vanilla/cache.rs

```diff
@@ -134,23 +156,99 @@ impl ParentCache {
+    pub fn open<H, G>(
+        len: u32,
+        cache_entries: u32,
+        graph: &StackedGraph<H, G>,
+        path: PathBuf,
+    ) -> Result<Self>
+    where
+        H: Hasher,
+        G: Graph<H> + ParameterSetMetadata + Send + Sync,
+    {
+        // Check if current entry is part of the official manifest.
+        // If not, we're dealing with some kind of test sector.
+        let (parent_cache_data, verify_cache, mut digest_hex) = match get_parent_cache_data(&path) {
+            None => {
+                info!("[open] Parent cache data is not supported in production");
+                (None, false, "".to_string())
+            }
+            Some(pcd) => (
+                Some(pcd),
+                settings::SETTINGS
+                    .lock()
+                    .expect("verify_cache settings lock failure")
+                    .verify_cache,
+                pcd.digest.clone(),
+            ),
+        };
+
+        info!(
+            "parent cache: opening {}, verify enabled: {}",
+            path.display(),
+            verify_cache
+        );
+
+        if verify_cache {
+            let parent_cache_data = parent_cache_data.expect("parent_cache_data failure");
 
-        let cache = CacheData::open(0, len, &path)?;
-        info!("parent cache: opened");
+            // Always check all of the data for integrity checks, even
+            // if we're only opening a portion of it.
+            let mut hasher = Sha256::new();
+            info!("[open] parent cache: calculating consistency digest");
+            let file = File::open(&path)?;
+            let data = unsafe {
+                memmap::MmapOptions::new()
+                    .map(&file)
+                    .with_context(|| format!("could not mmap path={}", path.display()))?
+            };
+            hasher.update(&data);
+            drop(data);
+
+            let hash = hasher.finalize();
+            info!("[open] parent cache: calculated consistency digest");
+
+            digest_hex = hash.iter().map(|x| format!("{:01$x}", x, 2)).collect();
+
+            trace!(
+                "[{}] Comparing {:?} to {:?}",
+                graph.size() * NODE_SIZE,
+                digest_hex,
+                parent_cache_data.digest
+            );
+            if digest_hex == parent_cache_data.digest {
+                info!("[open] parent cache: cache is verified!");
+            } else {
+                info!(
+                    "[!!!] Parent cache digest mismatch detected.  Regenerating {}",
+                    path.display()
+                );
+                ensure!(
+                    Self::generate(len, graph.size() as u32, graph, path.clone()).is_ok(),
+                    "Failed to generate parent cache"
+                );
+
+                // Note that if we wanted the user to manually terminate after repeated
+                // generation attemps, we could recursively return Self::open(...) here.
```

Comment:
We don't need to use recursion, or even necessarily loop, but if we have been asked to verify the cache, we *must* at least check it once after generating (and fail if we are not going to retry). Otherwise, those with bad hardware (e.g. RAM) who cannot generate a correct cache will not discover it. If `verify_cache` is specified, we *must not* open a bad cache because we didn't check it.

Reply:
I added the comment per your suggestion that maybe it's the right thing to keep generating and let the user terminate it after many attempts.  In either case, the graph is checked after generation and that check results in a failure.